File: Medium/238_ProductOfArrayExceptSelf/C++/Solution.cpp
#include <vector>

class Solution {
public:
    std::vector<int> productExceptSelf(std::vector<int>& nums) {
        std::vector<int> prefixes = {1}, postfixes = {1};
        int len = nums.size();
        for (int i = 0; i < len - 1; i++) {
            prefixes.emplace_back(nums[i] * prefixes[i]);
            postfixes.emplace_back(nums[len - 1 - i] * postfixes[i]);
        }      
        std::vector<int> answer;
        for (int i = 0; i < len; i++) answer.emplace_back(prefixes[i] * postfixes[postfixes.size() - 1 - i]);
        return answer;
    }
};

File: Medium/238_ProductOfArrayExceptSelf/Java/Solution.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] answer = new int[nums.length];
        
        // looping forward
        answer[0] = 1;
        for (int i = 1; i < nums.length; i++) answer[i] = answer[i - 1] * nums[i - 1];

        // looping backwards
        int x = 1;
        for (int i = nums.length - 1; i > -1; i--) {
            answer[i] *= x;
            x *= nums[i];
        }

        return answer;
    }
}